rclcpp_components/doc/architecture.rst

```diff
@@ -0,0 +1,263 @@
+
+Terminology
+-----------
+
+- Composable Node:
+  - An ``rclcpp`` based node that is setup so that it can be composed into a single process with other composable nodes at either runtime or programming time.
+- Node Component:
+  - Same thing as Composable Node, can be used interchangeably.
+- Composable Node Container:
+  - A program that can load and host composable nodes, it is often general purpose and has not a priori knowledge about the nodes it is instantiating and executing.
+- pluginlib:
+  - ROS package that provides mechanisms to create and register plugins based on user defined interface classes, and then dynamically lookup and load plugins at runtime.
+- Stand-alone Node Executable:
+  - Convenience executable that can be used to run a composable node as a single process or have it loaded dynamically in a remote composable node container.
+
+Theory of Operation
+-------------------
+
+The lowest level series of steps require to make a composable node are as follows:
+
+- Create a shared library in which one or more of your composable nodes are compiled.
+  - The nodes should also be registered as a ``pluginlib``-style plugin.
+  - The library should be created with CMake's ``add_library``
+  - The library should either explicitly be set to ``SHARED``.
+- Create or add to a ``plugindescription.xml`` file and register it so that ``pluginlib`` can find the node.
+- Create a stand-alone executable for each node.
+  - Which can either be used to run the node in its own process or to load it into an existing process and serving as a proxy while it is in the remote process.
+
+There are both C++ preprocessor and CMake macros within this package to assist with each of the above steps.
+Generally they can be applied in the following way:
+
+- Write files containing nodes, e.g. ``src/a.cpp``, ``src/b.cpp``, etc., and in each use the ``RCLCPP_COMPONENTS_REGISTER_NODE()`` preprocessor macro in the source code.
+- In CMake, create the shared library, e.g. ``add_library(my_library SHARED src/a.cpp src/b.cpp)``.
+  - Do any extra CMake steps for the library, e.g. ``target_link_libraries`` or ``target_include_directories``.
+- Describe each node in the library with the CMake using ``rclcpp_components_describe_node_component``.
+  - e.g. ``rclcpp_components_describe_node_component(my_library NAME MyNodeA TYPE my_cpp_namespace::MyNodeA DESCRIPTION "...")``
+- Register the library with ``pluginlib`` by calling ``rclcpp_components_register_node_component_library``.
+  - e.g. ``rclcpp_components_register_node_component_library(my_library GENERATE_PLUGIN_DESCRIPTION)``
+- Optionally, create a stand-alone executable for each node using ``rclcpp_components_create_executable_for_node``.
+  - e.g. ``rclcpp_components_create_executable_for_node(my_node_a_target NODE_COMPONENT_NAME MyNodeA)``
+  - An assumption is that package name is equal to ``PROJECT_NAME``, but can be overridden with the ``PACKAGE_NAME <package name>`` option.
```

Comment:
I think this set of steps is from an older version.  My understanding is that nowadays, we only need to call `rclcpp_components_register_nodes`, though I could be wrong.